### PR TITLE
Allow for custom css sheet overrides for reports

### PR DESCRIPTION
**Pull request by jchorl**
Adds the ability to specify CSS overrides. I know it's possible to add css to custom modules, but we'd like to do it for the simple/default reports, without specifying entirely new modules.

For example, I'd like to add css rules for the [css `print` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media). I will merge these with upstream. But for the short term, it's useful to just specify a bunch of overrides, and then get them merged. This will allow for e.g. beautiful print exports of multiqc reports (currently e.g. background colours don't show up). Let me know what you think!

- [x] This comment contains a description of changes (with reason)
- [x] `CHANGELOG.md` has been updated

I tried this out with a few configurations:

```console
$ cat sample.css 
.side-nav {
  width: 800px;
}
$ cat other_sample.css 
.side-nav h1 img {
  height: 200px;
}
$ cat multiqc_config.yaml 
custom_css_files:
  - sample.css
  - other_sample.css
```
Screenshot 2021-11-02 at 03-45-51 MultiQC Report

You can see the logo and sidebar are huge here.

Then I built/ran with cli flags:
```console
$ docker run -it --rm -v $(pwd):/work -w /work jchorl/multiqc --custom-css-files sample.css --custom-css-files other_sample.css .

  /// MultiQC 🔍 | v1.12.dev0

|           multiqc | Search path : /work
|         searching | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 3/3  
|    custom_content | plasmid_table: Found 2 samples (table)
|           multiqc | Compressing plot data
|           multiqc | Report      : multiqc_report.html
|           multiqc | Data        : multiqc_data
|           multiqc | MultiQC complete
 j@bike  ~/third-party/MultiQC/testdata   jchorl/customcss ±  ls multiqc_config.yaml
ls: cannot access 'multiqc_config.yaml': No such file or directory
```
Screenshot 2021-11-02 at 03-48-03 MultiQC Report

Finally, I removed the flags and config file:
Screenshot 2021-11-02 at 03-49-47 MultiQC Report

We're back to normal looking reports.

### CODE REVIEW

**Comment by jchorl** [comment]
@ewels any chance I could get you to take a look at this? Happy to revise as you see helpful.

File: multiqc/templates/default/includes.html [hunk 1]
```diff
@@ -32,6 +32,9 @@
     {{ include_file('assets/css/default_multiqc.css') }}
     {{ include_file('assets/css/jquery.toast.css') }}
 </style>
+{% for css_href in config.custom_css_files %}
+<style type="text/css">{{ include_file(css_href, None) }}</style>
```
Comment (ewels) [inline multiqc/templates/default/includes.html:36]: I forget, what is the second argument for? Is it needed?

File: multiqc/multiqc.py [hunk 2]
```diff
@@ -438,6 +449,8 @@ def run(
         config.exclude_modules = exclude
     if profile_runtime:
         config.profile_runtime = True
+    if custom_css_files is not None and len(custom_css_files) > 0:
```
Comment (ewels) [inline multiqc/multiqc.py:452]: I have a feeling that if a click argument accepts multiple, it returns an empty list of not specified. So you can probably ditch this line if that's true.

But worth checking as this is off the top of my head 😅

File: multiqc/multiqc.py [hunk 3]
```diff
@@ -176,6 +176,14 @@
 @click.option("-q", "--quiet", is_flag=True, help="Only show log warnings")
 @click.option("--profile-runtime", is_flag=True, help="Add analysis of how long MultiQC takes to run to the report")
 @click.option("--no-ansi", is_flag=True, help="Disable coloured log output")
+@click.option(
+    "--custom-css-files",
```
Comment (ewels) [inline multiqc/multiqc.py:180]: I don't think that you need this twice..

File: multiqc/multiqc.py [hunk 4]
```diff
@@ -176,6 +176,14 @@
...
+    "--custom-css-files",
+    "custom_css_files",
+    type=click.Path(exists=True, readable=True),
+    multiple=True,
+    help="Custom CSS files to add to the final report",
```
Comment (ewels) [inline multiqc/multiqc.py:185]: One per use of the flag, so I prefer singular here. 

```suggestion
    help="Custom CSS file to add to the final report",
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -30,6 +30,7 @@ Fixed logger bugs when calling `multiqc.run` multiple times by removing logging
   - Required a change to the `clean_s_name()` function arguments. All core MultiQC modules updated to reflect this.
   - Should be backwards compatible for custom modules. To adopt new behaviour, supply `f` instead of `f["root"]` as the second argument.
   - See the documenation for details: [Using log filenames as sample names](https://multiqc.info/docs/#using-log-filenames-as-sample-names) and [Custom sample names](https://multiqc.info/docs/#custom-sample-names).
+  - Added support for `--custom-css-files` / config `custom_css_files` option to include custom CSS in the final report
```
Comment (ewels) [inline CHANGELOG.md:33]:
```suggestion
  - Added support for `--custom-css-files` / `config.custom_css_files` option to include custom CSS in the final report
```

File: docs/config.md [hunk 6]
```diff
@@ -509,3 +509,16 @@ This approach is **not recommended if you have a very large number of samples**,
 produce a huge report file with all of the embedded plot data and crash your browser when opening it.
 If you are running MultiQC for the `multiqc_data` folder and never intend to look at the report, it
 speed things up though.
+
+## Custom CSS files
+
+MultiQC generates an html report. You can include custom CSS in your final report.
```
Comment (ewels) [inline docs/config.md:515]:
```suggestion
MultiQC generates HTML reports. You can include custom CSS in your final report if you wish.
```

File: docs/config.md [hunk 7]
```diff
@@ -509,3 +509,16 @@ This approach is **not recommended if you have a very large number of samples**,
...
+Simply add css files to the `custom_css_files` config option:
```
Comment (ewels) [inline docs/config.md:516]:
```suggestion
Simply add CSS files to the `custom_css_files` config option:
```

File: docs/config.md [hunk 8]
```diff
@@ -509,3 +509,16 @@ This approach is **not recommended if you have a very large number of samples**,
...
+
+```yaml
+custom_css_files:
+  - myfile.css
+```
+
+Or pass `--custom-css-files` (multiple params accepted) and multiqc will include
```
Comment (ewels) [inline docs/config.md:523]:
```suggestion
Or pass `--custom-css-files` (can be specified multiple times) and MultiQC will include
```

File: docs/config.md [hunk 9]
```diff
@@ -509,3 +509,16 @@ This approach is **not recommended if you have a very large number of samples**,
...
+them in the final report html.
```
Comment (ewels) [inline docs/config.md:524]:
```suggestion
them in the final report HTML.
```

**Review by ewels — APPROVED** [review]
Yeah, sorry - I've been on paternity leave and now working 50% so I'm very behind on MultiQC maintenance work.

Love the idea of this PR though, have done a quick read on my phone with one or two minor comments but happy to merge once resolved 👍🏻

Thanks for this!

File: multiqc/templates/default/includes.html (hunk 1 above)
Comment (jchorl) [reply multiqc/templates/default/includes.html:36]:
```
    def include_file(name, fdir=tmp_dir, b64=False):
        try:
            if fdir is None:
                fdir = ""
            if b64:
                with io.open(os.path.join(fdir, name), "rb") as f:
                    return base64.b64encode(f.read()).decode("utf-8")
            else:
                with io.open(os.path.join(fdir, name), "r", encoding="utf-8") as f:
                    return f.read()
```

It's the root dir, looks like. I suspect we can actually omit it. I just copy/pasta'd from right above it.

File: multiqc/multiqc.py (hunk 2 above)
Comment (jchorl) [reply multiqc/multiqc.py:452]: Will test!

File: multiqc/multiqc.py (hunk 3 above)
Comment (jchorl) [reply multiqc/multiqc.py:180]: You're probably right. Above:
```
@click.option(
    "--cl-config", "--cl_config", type=str, multiple=True, help="Specify MultiQC config YAML on the command line"
)
```

But lemme try taking it out :)

**Comment by jchorl** [comment]
Congrats on the pat leave :)

I implemented your suggestions. Thanks for working through.

If you have any additional cycles, I cut another 3 PRs. These 2 should be a slam dunk:
- https://github.com/ewels/MultiQC/pull/1575
- https://github.com/ewels/MultiQC/pull/1576

And this one's a tad hairier:
- https://github.com/ewels/MultiQC/pull/1577

**Comment by ewels** [comment]
Nice, thanks! Ready to merge apart from the failing CI tests that are due to changing CSPs. I never completely wrapped my head around this stuff. Are you ok to look into it and try to get the tests passing?

**Comment by ewels** [comment]
Ah crap, looks like that test is failing on all PRs. Ok may need some thought, I'd be nice to have some green ticks before merging if possible..

**Comment by jchorl** [comment]
Hey @ewels I actually fixed CI in https://github.com/ewels/MultiQC/pull/1578

Let's merge that, and then I can rebase the other PRs, get CI green, $$$?

**Comment by ewels** [comment]
Ah you're a hero! No need to rebase, can just rerun actions.

File: multiqc/multiqc.py (hunk 3 above)
Comment (ewels) [reply multiqc/multiqc.py]: Yeah, past-me probably didn't know better. Only needed when doing short form + long form if I remember correctly.

**Comment by jchorl** [comment]
Seems like actions run at the current sha, not rebasing on master. Should I just rebase and push?

**Comment by ewels** [comment]
Ok fab, thanks!

**Comment by ewels** [comment]
It's none of my business, but how come an engineer from Stripe is running MultiQC? Do you have some bioinformatics side project or are you using MultiQC for something totally different?

**Comment by jchorl** [comment]
> It's none of my business, but how come an engineer from Stripe is running MultiQC? Do you have some bioinformatics side project or are you using MultiQC for something totally different?

Nothing employer/work related - just nerding out on weekends, trying to learn :P

**Comment by ewels** [comment]
Awesome. Happy to have you here 😀

**Pull request merged**